File: common/python/system_config.py
"""Strict loader for the shared Zendure system configuration."""

from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any, Optional, Union
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError


DEFAULT_SYSTEM_CONFIG_PATH = Path(__file__).resolve().parents[1] / "config" / "system.json"


class SystemConfigError(ValueError):
    """Raised when the shared system configuration is missing or invalid."""


def load_system_config(config_path: Optional[Union[Path, str]] = None) -> dict[str, Any]:
    """Load, validate and normalize the shared system configuration."""

    path = Path(config_path) if config_path is not None else DEFAULT_SYSTEM_CONFIG_PATH
    if not path.is_file():
        raise SystemConfigError(f"System configuration file not found: {path}")

    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as error:
        raise SystemConfigError(f"Unable to read system configuration file: {path}") from error

    try:
        decoded = json.loads(raw)
    except json.JSONDecodeError as error:
        raise SystemConfigError(
            f"Invalid JSON in system configuration file {path}: "
            f"line {error.lineno} column {error.colno}"
        ) from error

    if not isinstance(decoded, dict):
        raise SystemConfigError("Expected an object at $.")
    return validate_system_config(decoded)


def validate_system_config(config: dict[str, Any]) -> dict[str, Any]:
    """Validate and normalize an already decoded system configuration."""

    _assert_exact_keys(
        config,
        {"schemaVersion", "battery", "installation", "priceConversion"},
        "$",
    )

    schema_version = _require_integer(config["schemaVersion"], "$.schemaVersion", 1, 1)
    battery = _require_object(config["battery"], "$.battery")
    installation = _require_object(config["installation"], "$.installation")
    price_conversion = _require_object(config["priceConversion"], "$.priceConversion")

    _assert_exact_keys(
        battery,
        {"capacityWh", "minChargePercent", "maxChargePercent"},
        "$.battery",
    )
    capacity_wh = _require_integer(battery["capacityWh"], "$.battery.capacityWh", 1)
    min_charge_percent = _require_integer(
        battery["minChargePercent"], "$.battery.minChargePercent", 0, 99
    )
    max_charge_percent = _require_integer(
        battery["maxChargePercent"], "$.battery.maxChargePercent", 1, 100
    )
    if min_charge_percent >= max_charge_percent:
        raise SystemConfigError(
            "$.battery.minChargePercent must be lower than $.battery.maxChargePercent."
        )

    _assert_exact_keys(
        installation,
        {"name", "latitude", "longitude", "timezone"},
        "$.installation",
    )
    name = _require_non_blank_string(installation["name"], "$.installation.name")
    latitude = _require_number(installation["latitude"], "$.installation.latitude", -90.0, 90.0)
    longitude = _require_number(
        installation["longitude"], "$.installation.longitude", -180.0, 180.0
    )
    timezone = _require_non_blank_string(installation["timezone"], "$.installation.timezone")
    try:
        ZoneInfo(timezone)
    except (ZoneInfoNotFoundError, ValueError) as error:
        raise SystemConfigError("$.installation.timezone is not a recognized IANA timezone.") from error

    _assert_exact_keys(
        price_conversion,
        {
            "supplierMarkupEurPerKwh",
            "energyTaxEurPerKwh",
            "vatMultiplier",
            "consumerPrecision",
            "spotPrecision",
        },
        "$.priceConversion",
    )
    supplier_markup = _require_number(
        price_conversion["supplierMarkupEurPerKwh"],
        "$.priceConversion.supplierMarkupEurPerKwh",
        0.0,
    )
    energy_tax = _require_number(
        price_conversion["energyTaxEurPerKwh"],
        "$.priceConversion.energyTaxEurPerKwh",
        0.0,
    )
    vat_multiplier = _require_number(
        price_conversion["vatMultiplier"],
        "$.priceConversion.vatMultiplier",
        0.0,
        exclusive_minimum=True,
    )
    consumer_precision = _require_integer(
        price_conversion["consumerPrecision"],
        "$.priceConversion.consumerPrecision",
        0,
        12,
    )
    spot_precision = _require_integer(
        price_conversion["spotPrecision"],
        "$.priceConversion.spotPrecision",
        0,
        12,
    )

    return {
        "schemaVersion": schema_version,
        "battery": {
            "capacityWh": capacity_wh,
            "minChargePercent": min_charge_percent,
            "maxChargePercent": max_charge_percent,
        },
        "installation": {
            "name": name,
            "latitude": latitude,
            "longitude": longitude,
            "timezone": timezone,
        },
        "priceConversion": {
            "supplierMarkupEurPerKwh": supplier_markup,
            "energyTaxEurPerKwh": energy_tax,
            "vatMultiplier": vat_multiplier,
            "consumerPrecision": consumer_precision,
            "spotPrecision": spot_precision,
        },
    }


def _require_object(value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise SystemConfigError(f"Expected an object at {path}.")
    return value


def _assert_exact_keys(value: dict[str, Any], expected: set[str], path: str) -> None:
    actual = set(value)
    missing = sorted(expected - actual)
    unknown = sorted(actual - expected)
    if not missing and not unknown:
        return

    details = []
    if missing:
        details.append(f"missing: {', '.join(missing)}")
    if unknown:
        details.append(f"unknown: {', '.join(unknown)}")
    raise SystemConfigError(f"Invalid properties at {path} ({'; '.join(details)}).")


def _require_integer(value: Any, path: str, minimum: int, maximum: Optional[int] = None) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise SystemConfigError(f"{path} must be an integer.")
    if value < minimum:
        raise SystemConfigError(f"{path} must be at least {minimum}.")
    if maximum is not None and value > maximum:
        raise SystemConfigError(f"{path} must be at most {maximum}.")
    return value


def _require_number(
    value: Any,
    path: str,
    minimum: float,
    maximum: Optional[float] = None,
    *,
    exclusive_minimum: bool = False,
) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise SystemConfigError(f"{path} must be a number.")

    number = float(value)
    if not math.isfinite(number):
        raise SystemConfigError(f"{path} must be a finite number.")
    if (exclusive_minimum and number <= minimum) or (not exclusive_minimum and number < minimum):
        comparison = "greater than" if exclusive_minimum else "at least"
        raise SystemConfigError(f"{path} must be {comparison} {_format_number(minimum)}.")
    if maximum is not None and number > maximum:
        raise SystemConfigError(f"{path} must be at most {_format_number(maximum)}.")
    return number


def _require_non_blank_string(value: Any, path: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise SystemConfigError(f"{path} must be a non-blank string.")
    return value


def _format_number(value: float) -> str:
    return f"{value:.12f}".rstrip("0").rstrip(".")

File: common/php/system_config.php
<?php

declare(strict_types=1);

final class SystemConfigException extends RuntimeException
{
}

function systemConfigDefaultPath(): string
{
    return dirname(__DIR__) . '/config/system.json';
}

/**
 * Load and validate the shared system configuration.
 *
 * @return array{
 *   schemaVersion: int,
 *   battery: array{
 *     capacityWh: int,
 *     minChargePercent: int,
 *     maxChargePercent: int,
 *     efficiency: float,
 *     maxChargePowerW: int,
 *     maxDischargePowerW: int
 *   },
 *   forecast: array{defaultHouseholdUsageWByHour: list<int>},
 *   schedule: array{minPowerW: int, maxPowerW: int, powerStepW: int},
 *   installation: array{name: string, latitude: float, longitude: float, timezone: string},
 *   priceConversion: array{
 *     supplierMarkupEurPerKwh: float,
 *     energyTaxEurPerKwh: float,
 *     vatMultiplier: float,
 *     consumerPrecision: int,
 *     spotPrecision: int
 *   }
 * }
 */
function loadSystemConfig(?string $configPath = null): array
{
    $path = $configPath ?? systemConfigDefaultPath();
    if (!is_file($path)) {
        throw new SystemConfigException('System configuration file not found: ' . $path);
    }

    $raw = @file_get_contents($path);
    if ($raw === false) {
        throw new SystemConfigException('Unable to read system configuration file: ' . $path);
    }

    try {
        $rootValue = json_decode($raw, false, 512, JSON_THROW_ON_ERROR);
        $decoded = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
    } catch (JsonException $error) {
        throw new SystemConfigException(
            'Invalid JSON in system configuration file ' . $path . ': ' . $error->getMessage(),
            0,
            $error
        );
    }

    if (!$rootValue instanceof stdClass || !is_array($decoded)) {
        throw new SystemConfigException('Expected an object at $.');
    }

    return validateSystemConfig($decoded);
}

/**
 * Validate and normalize an already decoded system configuration.
 *
 * @param array<mixed> $config
 * @return array<string, mixed>
 */
function validateSystemConfig(array $config): array
{
    systemConfigAssertExactKeys(
        $config,
        ['schemaVersion', 'battery', 'forecast', 'schedule', 'installation', 'priceConversion'],
        '$'
    );

    $schemaVersion = systemConfigRequireInteger($config['schemaVersion'], '$.schemaVersion', 1, 1);
    $battery = systemConfigRequireObject($config['battery'], '$.battery');
    $forecast = systemConfigRequireObject($config['forecast'], '$.forecast');
    $schedule = systemConfigRequireObject($config['schedule'], '$.schedule');
    $installation = systemConfigRequireObject($config['installation'], '$.installation');
    $priceConversion = systemConfigRequireObject($config['priceConversion'], '$.priceConversion');

    systemConfigAssertExactKeys(
        $battery,
        [
            'capacityWh',
            'minChargePercent',
            'maxChargePercent',
            'efficiency',
            'maxChargePowerW',
            'maxDischargePowerW',
        ],
        '$.battery'
    );
    $capacityWh = systemConfigRequireInteger($battery['capacityWh'], '$.battery.capacityWh', 1);
    $minChargePercent = systemConfigRequireInteger(
        $battery['minChargePercent'],
        '$.battery.minChargePercent',
        0,
        99
    );
    $maxChargePercent = systemConfigRequireInteger(
        $battery['maxChargePercent'],
        '$.battery.maxChargePercent',
        1,
        100
    );
    if ($minChargePercent >= $maxChargePercent) {
        throw new SystemConfigException(
            '$.battery.minChargePercent must be lower than $.battery.maxChargePercent.'
        );
    }
    $efficiency = systemConfigRequireNumber($battery['efficiency'], '$.battery.efficiency', 0.0, 1.0, true);
    $maxChargePowerW = systemConfigRequireInteger(
        $battery['maxChargePowerW'],
        '$.battery.maxChargePowerW',
        1
    );
    $maxDischargePowerW = systemConfigRequireInteger(
        $battery['maxDischargePowerW'],
        '$.battery.maxDischargePowerW',
        1
    );

    systemConfigAssertExactKeys(
        $forecast,
        ['defaultHouseholdUsageWByHour'],
        '$.forecast'
    );
    $defaultHouseholdUsage = systemConfigRequireList(
        $forecast['defaultHouseholdUsageWByHour'],
        '$.forecast.defaultHouseholdUsageWByHour',
        24
    );
    $defaultHouseholdUsageWByHour = [];
    foreach ($defaultHouseholdUsage as $hour => $usageW) {
        $defaultHouseholdUsageWByHour[] = systemConfigRequireInteger(
            $usageW,
            '$.forecast.defaultHouseholdUsageWByHour[' . $hour . ']',
            0
        );
    }

    systemConfigAssertExactKeys(
        $schedule,
        ['minPowerW', 'maxPowerW', 'powerStepW'],
        '$.schedule'
    );
    $minPowerW = systemConfigRequireInteger($schedule['minPowerW'], '$.schedule.minPowerW', PHP_INT_MIN, 0);
    $maxPowerW = systemConfigRequireInteger($schedule['maxPowerW'], '$.schedule.maxPowerW', 0);
    $powerStepW = systemConfigRequireInteger($schedule['powerStepW'], '$.schedule.powerStepW', 1);
    if ($minPowerW >= $maxPowerW) {
        throw new SystemConfigException('$.schedule.minPowerW must be lower than $.schedule.maxPowerW.');
    }

    systemConfigAssertExactKeys(
        $installation,
        ['name', 'latitude', 'longitude', 'timezone'],
        '$.installation'
    );
    $name = systemConfigRequireNonBlankString($installation['name'], '$.installation.name');
    $latitude = systemConfigRequireNumber($installation['latitude'], '$.installation.latitude', -90.0, 90.0);
    $longitude = systemConfigRequireNumber($installation['longitude'], '$.installation.longitude', -180.0, 180.0);
    $timezone = systemConfigRequireNonBlankString($installation['timezone'], '$.installation.timezone');
    if (!in_array($timezone, DateTimeZone::listIdentifiers(), true)) {
        throw new SystemConfigException('$.installation.timezone is not a recognized IANA timezone.');
    }

    systemConfigAssertExactKeys(
        $priceConversion,
        [
            'supplierMarkupEurPerKwh',
            'energyTaxEurPerKwh',
            'vatMultiplier',
            'consumerPrecision',
            'spotPrecision',
        ],
        '$.priceConversion'
    );
    $supplierMarkup = systemConfigRequireNumber(
        $priceConversion['supplierMarkupEurPerKwh'],
        '$.priceConversion.supplierMarkupEurPerKwh',
        0.0
    );
    $energyTax = systemConfigRequireNumber(
        $priceConversion['energyTaxEurPerKwh'],
        '$.priceConversion.energyTaxEurPerKwh',
        0.0
    );
    $vatMultiplier = systemConfigRequireNumber(
        $priceConversion['vatMultiplier'],
        '$.priceConversion.vatMultiplier',
        0.0,
        null,
        true
    );
    $consumerPrecision = systemConfigRequireInteger(
        $priceConversion['consumerPrecision'],
        '$.priceConversion.consumerPrecision',
        0,
        12
    );
    $spotPrecision = systemConfigRequireInteger(
        $priceConversion['spotPrecision'],
        '$.priceConversion.spotPrecision',
        0,
        12
    );

    return [
        'schemaVersion' => $schemaVersion,
        'battery' => [
            'capacityWh' => $capacityWh,
            'minChargePercent' => $minChargePercent,
            'maxChargePercent' => $maxChargePercent,
            'efficiency' => $efficiency,
            'maxChargePowerW' => $maxChargePowerW,
            'maxDischargePowerW' => $maxDischargePowerW,
        ],
        'forecast' => [
            'defaultHouseholdUsageWByHour' => $defaultHouseholdUsageWByHour,
        ],
        'schedule' => [
            'minPowerW' => $minPowerW,
            'maxPowerW' => $maxPowerW,
            'powerStepW' => $powerStepW,
        ],
        'installation' => [
            'name' => $name,
            'latitude' => $latitude,
            'longitude' => $longitude,
            'timezone' => $timezone,
        ],
        'priceConversion' => [
            'supplierMarkupEurPerKwh' => $supplierMarkup,
            'energyTaxEurPerKwh' => $energyTax,
            'vatMultiplier' => $vatMultiplier,
            'consumerPrecision' => $consumerPrecision,
            'spotPrecision' => $spotPrecision,
        ],
    ];
}

/** @param mixed $value @return array<mixed> */
function systemConfigRequireObject($value, string $path): array
{
    if (!is_array($value)) {
        throw new SystemConfigException('Expected an object at ' . $path . '.');
    }
    return $value;
}

/** @param mixed $value @return list<mixed> */
function systemConfigRequireList($value, string $path, int $length): array
{
    if (!is_array($value) || !array_is_list($value)) {
        throw new SystemConfigException('Expected an array at ' . $path . '.');
    }
    if (count($value) !== $length) {
        throw new SystemConfigException($path . ' must contain exactly ' . $length . ' items.');
    }
    return $value;
}

/** @param array<mixed> $value @param list<string> $expected */
function systemConfigAssertExactKeys(array $value, array $expected, string $path): void
{
    $actual = array_keys($value);
    $missing = array_values(array_diff($expected, $actual));
    $unknown = array_values(array_diff($actual, $expected));
    if ($missing === [] && $unknown === []) {
        return;
    }

    $details = [];
    if ($missing !== []) {
        $details[] = 'missing: ' . implode(', ', $missing);
    }
    if ($unknown !== []) {
        $details[] = 'unknown: ' . implode(', ', array_map('strval', $unknown));
    }
    throw new SystemConfigException('Invalid properties at ' . $path . ' (' . implode('; ', $details) . ').');
}

/** @param mixed $value */
function systemConfigRequireInteger($value, string $path, int $minimum, ?int $maximum = null): int
{
    if (!is_int($value)) {
        throw new SystemConfigException($path . ' must be an integer.');
    }
    if ($value < $minimum) {
        throw new SystemConfigException($path . ' must be at least ' . $minimum . '.');
    }
    if ($maximum !== null && $value > $maximum) {
        throw new SystemConfigException($path . ' must be at most ' . $maximum . '.');
    }
    return $value;
}

/** @param mixed $value */
function systemConfigRequireNumber(
    $value,
    string $path,
    float $minimum,
    ?float $maximum = null,
    bool $exclusiveMinimum = false
): float {
    if (!is_int($value) && !is_float($value)) {
        throw new SystemConfigException($path . ' must be a number.');
    }

    $number = (float)$value;
    if (!is_finite($number)) {
        throw new SystemConfigException($path . ' must be a finite number.');
    }
    if (($exclusiveMinimum && $number <= $minimum) || (!$exclusiveMinimum && $number < $minimum)) {
        $comparison = $exclusiveMinimum ? 'greater than ' : 'at least ';
        throw new SystemConfigException($path . ' must be ' . $comparison . systemConfigFormatNumber($minimum) . '.');
    }
    if ($maximum !== null && $number > $maximum) {
        throw new SystemConfigException($path . ' must be at most ' . systemConfigFormatNumber($maximum) . '.');
    }
    return $number;
}

/** @param mixed $value */
function systemConfigRequireNonBlankString($value, string $path): string
{
    if (!is_string($value) || trim($value) === '') {
        throw new SystemConfigException($path . ' must be a non-blank string.');
    }
    return $value;
}

function systemConfigFormatNumber(float $value): string
{
    return rtrim(rtrim(number_format($value, 12, '.', ''), '0'), '.');
}
